granulate_utils/linux/cgroups/cpu_controller.py

```diff
@@ -0,0 +1,110 @@
+    def get_cpu_limit_params(self) -> CpuLimitParams:
+        pass
+
+    @abstractmethod
+    def set_cpu_limit_quota(self, quota: str) -> None:
+        pass
+
+
+class CpuControllerV1(CpuControllerInterface):
+    cpu_stat = "cpu.stat"
+    period = "cpu.cfs_period_us"
+    quota = "cpu.cfs_quota_us"
+
+    def get_cpu_limit_period(self) -> int:
+        return int(self.cgroup.read_from_interface_file(self.period))
+
+    def get_cpu_limit_quota(self) -> int:
+        return int(self.cgroup.read_from_interface_file(self.quota))
+
+    def get_cpu_limit_params(self) -> CpuLimitParams:
+        return CpuLimitParams(self.get_cpu_limit_period(), self.get_cpu_limit_quota())
+
+    def set_cpu_limit_quota(self, quota: str) -> None:
+        self.cgroup.write_to_interface_file(self.quota, quota)
+
+
+class CpuControllerV2(CpuControllerInterface):
+    UNBOUNDED_QUOTA_VALUE = "max"
+    cpu_stat = "cpu.stat"
+    cpu_limit_file = "cpu.max"
+
+    def get_cpu_limit_period(self) -> int:
+        return self.get_cpu_limit_params().period
+
+    def get_cpu_limit_quota(self) -> int:
+        return self.get_cpu_limit_params().quota
+
+    def get_cpu_limit_params(self) -> CpuLimitParams:
+        cpu_limit = self.cgroup.read_from_interface_file(self.cpu_limit_file).split()
+        return CpuLimitParams(
+            period=int(cpu_limit[1]), quota=int(cpu_limit[0]) if cpu_limit[0] != self.UNBOUNDED_QUOTA_VALUE else -1
+        )
+
+    def set_cpu_limit_quota(self, quota: str) -> None:
+        period = self.get_cpu_limit_period()
+        if quota == "-1":
+            quota = self.UNBOUNDED_QUOTA_VALUE
+        self.cgroup.write_to_interface_file(self.cpu_limit_file, f"{quota} {period}")
+
+
+class CpuController(BaseController):
+    controller = "cpu"
+
+    def __init__(self, cgroup: Optional[Union[Path, CgroupCore]] = None):
+        super().__init__(cgroup)
+        if isinstance(self.cgroup, CgroupCoreV1):
+            self.controller_interface: CpuControllerInterface = CpuControllerV1(self.cgroup)
+        elif isinstance(self.cgroup, CgroupCoreV2):
+            self.controller_interface = CpuControllerV2(self.cgroup)
+
+    def set_cpu_limit_cores(self, cores: float) -> None:
+        period = self.controller_interface.get_cpu_limit_period()
+        self.controller_interface.set_cpu_limit_quota(str(int(period * cores)))
+
+    def get_cpu_limit_cores(self) -> float:
+        cpu_limit_params = self.controller_interface.get_cpu_limit_params()
+        # if quota is set to -1 it means this cgroup is unlimited
+        return cpu_limit_params.quota / cpu_limit_params.period if cpu_limit_params.quota != -1 else -1.0
+
+    def reset_cpu_limit(self) -> None:
+        self.controller_interface.set_cpu_limit_quota("-1")
+
+    def get_stat(self) -> Dict[str, int]:
+        stat_text = self.read_from_interface_file(self.controller_interface.cpu_stat)
+        return {line.split()[0]: int(line.split()[1]) for line in stat_text.splitlines()}
```

Comment:
Did you consider this class hierarchy (instead of your diamond hierarchy where D contains B C who extend A) - the methods here go to A, then B C will have them through inheritance, and just have a function D which gets the path and creates the appropriate object?

Reply:
That's what I did at first actually.
I changed it mainly because of the function D, which needs to be duplicated in each controller type (cpu, memory...) and I didn't find a nice way to share this code between them.
The way it is now I only need `base_controller.get_cgroup_in_hierarchy` and the user just needs to call:
```
memory_controller = MemoryController.get_cgroup_in_hierarchy(cgroup_name='blabla')
```
Anyway, I think I like it better this way but I do see the benefit of the other way so if you want I don't mind changing.

Reply:
The benefit is less indirections - no more classes containing classes, just classes extending classes, which is simpler to read & follow.